src/NServiceBus.Core.Tests/IdGeneration/IdGenerationTests.cs

```diff
@@ -0,0 +1,117 @@
+            var improvedCombGuid = LocalCopyOfCombGuid.GenerateImproved(guid, now);
+
+            Assert.AreEqual(oldCombGuid, improvedCombGuid);
+        }
+
+        static readonly Random Random = new Random();
+
+        static IEnumerable<object[]> RandomizedInput
+        {
+            get
+            {
+                for (int i = 0; i < 20; i++)
+                {
+                    yield return new object[]
+                    {
+                        Guid.NewGuid(),
+                        DateTime.UtcNow.AddDays(Random.Next(0, 30)).AddHours(Random.Next(0, 24)).AddMinutes(Random.Next(0, 60)).AddSeconds(Random.Next(0, 60))
+                    };
+                }
+            }
+        }
+
+        class LocalCopyOfCombGuid
+        {
+            // Only slightly modified copy of the original algorithm
+            public static Guid GenerateOld(Guid input, DateTime nowInput)
+            {
+                var guidArray = input.ToByteArray();
+
+                var now = nowInput; // Internal use, no need for DateTimeOffset
+
+                // Get the days and milliseconds which will be used to build the byte string
+                var days = new TimeSpan(now.Ticks - BaseDateTicks);
+                var timeOfDay = now.TimeOfDay;
+
+                // Convert to a byte array
+                // Note that SQL Server is accurate to 1/300th of a millisecond so we divide by 3.333333
+                var daysArray = BitConverter.GetBytes(days.Days);
+                var millisecondArray = BitConverter.GetBytes((long)(timeOfDay.TotalMilliseconds / 3.333333));
+
+                // Reverse the bytes to match SQL Servers ordering
+                Array.Reverse(daysArray);
+                Array.Reverse(millisecondArray);
+
+                // Copy the bytes into the guid
+                Array.Copy(daysArray, daysArray.Length - 2, guidArray, guidArray.Length - 6, 2);
+                Array.Copy(millisecondArray, millisecondArray.Length - 4, guidArray, guidArray.Length - 4, 4);
+
+                return new Guid(guidArray);
+            }
+
+            public static Guid GenerateImproved(Guid input, DateTime nowInput)
+            {
+                var newGuid = input;
+                Span<byte> guidArray = stackalloc byte[16];
+#if NETCOREAPP
+                if (!newGuid.TryWriteBytes(guidArray))
+#elif NETFRAMEWORK
+                if (!MemoryMarshal.TryWrite(guidArray, ref newGuid))
+#endif
+                {
+
+                    guidArray = newGuid.ToByteArray();
+                }
+
+                var now = nowInput; // Internal use, no need for DateTimeOffset
+
+                // Get the days and milliseconds which will be used to build the byte string
+                var days = new TimeSpan(now.Ticks - BaseDateTicks);
+                var timeOfDay = now.TimeOfDay;
+
+                // Convert to a byte array
+                Span<byte> daysArray = stackalloc byte[sizeof(int)];
```

Comment:
Technically, we could also go with one scratch array of the size of a long and use that to write things into or come up with a clever way to directly write the bytes necessary into `guidArray` but I figured that might be too clever and make the code unnecessary complex